### ISSUE TITLE
Could IExecutorService threads have their thread name altered

### ISSUE THREAD

**Issue by StanAccy** [hazelcast/hazelcast#6600]
Currently, all threads related to IExecutorService instances are allocated from the "cached" thread pool and are named accordingly: cached1..cachedN.

Would it be possible to append/insert the IExecutorService name to the created thread so its easier to see/diagnose in the debugger/profiler.

Eg.  IExecutorService.setName("foo");  would result in:
cached_foo_1
cached_foo_2
..
cached_foo_N

**Comment by pveentjer** [comment]
I never understood why a cached threadpool is being used here. That is the cause of the names of the threads being cached instead of concrete. I'll make a pr for it; I'm curious about the reasons behind this choice.

**Comment by StanAccy** [comment]
Great, thanks.  Perhaps the thread names could be execsrv-{IES_NAME}-1 (or something along those lines)

**Comment by pveentjer** [comment]
When I name the threadpool 'foo' this is how the thread is named

![screenshot_2015-10-29_23-25-56](https://cloud.githubusercontent.com/assets/105243/10832496/71452bf6-7e94-11e5-8d0d-96e9fa81dc1d.png)

**Comment by StanAccy** [comment]
The threadpool, yes.  But we can only name the IExecutorService - that's the name I think would be helpful.

**Comment by pveentjer** [comment]
But that is the name you use to retrieve the threapool from hz.

IExecutorService executor = hz.getExecutorService("thisismycoolexecutor")

**Comment by StanAccy** [comment]
When I name the IExecutorService "foo", I just get a series of threads named cached1..cachedN where N is the IExecutorService.poolSize

**Comment by pveentjer** [comment]
yes. But with my fix that should be addressed. If you name your IExecutorService foo, your threads will be called something with foo.

**Comment by StanAccy** [comment]
That sounds perfect.

**Comment by pveentjer** [comment]
https://github.com/hazelcast/hazelcast/pull/6601

**Comment by jerrinot** [comment]
You probably do not want to create a new thread pool for each IExecutorService - that's a big change when compared to the current implementation with negative side-effects. 

I understand descriptive thread names are useful during troubleshooting and for operation monitoring in general. I guess the way to move forward is to (temporary) rename the thread when task is about to be executed. This is a nice place: https://github.com/jerrinot/hazelcast/blob/d0fcf0bb8da8265c41458b3c6f4be023cabe5397/hazelcast/src/main/java/com/hazelcast/executor/impl/DistributedExecutorService.java#L205-L205

**Comment by pertsodian** [comment]
Hello, I can pick this one up. Looking to:
* add another static method in [ThreadUtil](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/util/ThreadUtil.java) to calculate the new thread name (assuming they all used createThreadPoolName previously)
* set and unset in [DistributedExecutorService](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/executor/impl/DistributedExecutorService.java) before and after executing the task as per @jerrinot's suggestion

**Comment by Holmistr** [comment]
Hi @pertsodian , and the issue is yours! :tada: I will get some of our experts to confirm your idea :point_up: or provide any further guidance. Looking forward to your contribution!

**Comment by pertsodian** [comment]
Hello, I raised #16695. Please take a look and let me know

**Comment by Holmistr** [comment]
Great work @pertsodian ! We're searching for reviewers as we speak :)

**Comment by mmedenjak** [comment]
Conclusion from PR - there is no good way in how the thread name can be changed "on-the-fly" as various profilers and tools show different results. As a result, the only approach we can think now is adding configuration support for creating concrete executors, all with different thread names. Unfortunately, this is a laborious task with questionable benefits. Allowing users to create threads easily may backfire as the system will be under increased load and the additional configuration part will make using hazelcast harder. So we have to re-evaluate the interest the community has in this approach.

**Comment by Holmistr** [comment]
Removing the "good first issue" label as, as explained above, there's no good way to do this. We will consider to focus on this issue if there's a bigger community interest.
